### ISSUE TITLE
Get the table schema as json?

### ISSUE THREAD

**Issue by eitsupi** [apache/arrow#13803]
I wanted to extract the schema of the table as json, but I noticed that the Python and R packages apparently only output the schema as a single string.
So I had to parse a string, but if a method was implemented that would output a dictionary or list from the schema, I would use that.

**Comment by nealrichardson** [comment]
The trouble is: what do you expect to do with this JSON? Round-trip and re-create the Schema from the JSON? That gets tricky with things like nested types, extension types, metadata, etc. It really becomes an Arrow specification question, not a JSON serialization problem. As such, if you want to pursue it, you should raise it on the dev mailing list. I seem to recall this idea coming up in the past and it not getting much support, but maybe I misremember, or maybe the times are different. 

Naively, in R you can get a JSON string version of the Schema with some manipulation. Here's an ugly base-R one-liner, just to illustrate:

```
> s
Schema
name: string
height: int32
mass: double
hair_color: string
skin_color: string
eye_color: string
birth_year: double
sex: string
gender: string
homeworld: string
species: string
films: list<item: string>
vehicles: list<item: string>
starships: list<item: string>

> cat(jsonlite::toJSON(setNames(lapply(s, function (x) x$type$ToString()), s$names), pretty = TRUE, auto_unbox = TRUE))
{
  "name": "string",
  "height": "int32",
  "mass": "double",
  "hair_color": "string",
  "skin_color": "string",
  "eye_color": "string",
  "birth_year": "double",
  "sex": "string",
  "gender": "string",
  "homeworld": "string",
  "species": "string",
  "films": "list<item: string>",
  "vehicles": "list<item: string>",
  "starships": "list<item: string>"
}
```

But if you wanted to create a Schema from that, you're still in the world of parsing things like `"list<item: string>"`.

On the R side, there is this nice code generating utility that Romain added recently:

```
> s$code()
schema(name = utf8(), height = int32(), mass = float64(), hair_color = utf8(), 
    skin_color = utf8(), eye_color = utf8(), birth_year = float64(), 
    sex = utf8(), gender = utf8(), homeworld = utf8(), species = utf8(), 
    films = list_of(utf8()), vehicles = list_of(utf8()), starships = list_of(utf8()))
```

**Comment by eitsupi** [comment]
Thanks for your reply.
My intended use is to recreate the schema from json and to do a simple validation of the table by checking the json.

I believe python currently allows we to create a schema from a json file. Like that:

```json
{
    "col1": "int32"
}
```

```python
>>> import pyarrow as pa
>>> import json
>>> with open("test.json") as f:
...     d = json.load(f)
... 
>>> pa.schema(d)
col1: int32
```

> On the R side, there is this nice code generating utility that Romain added recently:

I did not know this method existed because it was not mentioned in the R documentation.
I think this is pretty close to what I want.

**Comment by nealrichardson** [comment]
> Thanks for your reply. My intended uses are
> 
> 1. Recreate the schema from json files.
> 2. Simple validation of the table by checking json files.
> 
> I believe python currently allows we to create a schema from a json file. Like that:
> 
> ```json
> {
>     "col1": "int32"
> }
> ```
> 
> ```python
> >>> import pyarrow as pa
> >>> import json
> >>> with open("test.json") as f:
> ...     d = json.load(f)
> ... 
> >>> pa.schema(d)
> col1: int32
> ```
> 

Making a schema from a list like that nearly works in R, just would need to add support for accepting string type names instead of type instances:

```
> schema(list(col1 = "int32"))
Error: a must be a DataType, not character
> schema(list(col1 = int32()))
Schema
col1: int32
```

But you still have the problem with nested/list/dictionary types, which would require string parsing. Those don't work in pyarrow either:

```
>>> import pyarrow as pa
>>> pa.schema({"col1": "int32"})
col1: int32
>>> pa.schema({"col1": "dictionary<values=string, indices=int32>"})
Traceback (most recent call last):
  File "pyarrow/types.pxi", line 3171, in pyarrow.lib.type_for_alias
KeyError: 'dictionary<values=string, indices=int32>'
```

> > On the R side, there is this nice code generating utility that Romain added recently:
> 
> I did not know this method existed because it was not mentioned in the R documentation. I think this is pretty close to what I want.

Would you like to submit a PR to improve those docs? 😉  🙏

**Comment by eitsupi** [comment]
> Would you like to submit a PR to improve those docs? 😉 🙏

Yes, but apparently there are other undocumented methods and I need to understand them as well.

**Comment by wjones127** [comment]
FWIW, there is already an implementation of JSON roundtripping of schemas in arrow-rs. So if we did want to standardize, could simply propose using their format.

https://github.com/apache/arrow-rs/blob/27f4762c8794ef1c5d042933562185980eb85ae5/arrow/src/datatypes/datatype.rs#L785

**Comment by eitsupi** [comment]
> FWIW, there is already an implementation of JSON roundtripping of schemas in arrow-rs. So if we did want to standardize, could simply propose using their format.
> 
> [apache/arrow-rs@`27f4762`/arrow/src/datatypes/datatype.rs#L785](https://github.com/apache/arrow-rs/blob/27f4762c8794ef1c5d042933562185980eb85ae5/arrow/src/datatypes/datatype.rs#L785)

Thank you for letting me know about this!
This seems very promising to me, but why was it only implemented in Rust 4 years ago? (ARROW-2385, #1829)

**Comment by emkornfield** [comment]
If we were to implement this, I still think the right way of approaching it would be use Flatbuffers JSON representation for it.  Someone tried to get a PR going but ran into some blockers that I didn't get time to look at: https://github.com/apache/arrow/pull/7110

**Comment by wjones127** [comment]
> This seems very promising to me, but why was it only implemented in Rust 4 years ago? ([ARROW-2385](https://issues.apache.org/jira/browse/ARROW-2385), https://github.com/apache/arrow/pull/1829)

Ah it was actually specifically only for integration tests, but accidentally made public. It was just recently made private again in arrow-rs 24.

**Comment by kentkwu** [comment]
Recently there has been additional interest in this idea. I've started a discussion on the mailing list for what a proposed JSON representation might look like.
- https://lists.apache.org/thread/rbqmskk3omdfdhmsmtgz7v0vhmppgyw1

If this is interesting to you, please take a look and reply on the thread.
